### PR TITLE
[ADD] l10n_ar: Liquido Product document management

### PR DESCRIPTION
**Pull request by zaoral**
* Add refunds with liquido docs
* Add Is AFIP POS field (l10n_ar_is_pos)
* Adapt letter mapping

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/l10n_ar/models/account_move.py [hunk 1]
```diff
@@ -34,6 +34,17 @@ def _l10n_ar_get_document_number_parts(self, document_number, document_type_code
     l10n_ar_afip_service_start = fields.Date(string='AFIP Service Start Date', readonly=True, states={'draft': [('readonly', False)]})
     l10n_ar_afip_service_end = fields.Date(string='AFIP Service End Date', readonly=True, states={'draft': [('readonly', False)]})
 
+    def _is_manual_document_number(self):
+        res = super()._is_manual_document_number()
+        # NOTE: There is a corner case where 2 sales documents can have the same number for the same DOC from a different vendor,
+        # in that case, the user can create a new Sales Liquido Producto Journal
+
+        # When issuer is supplie the numbering works opposite (supplier enumerate the invoices, and the customer encode the bill)
+        if self.country_code == 'AR' and self.l10n_latam_use_documents and ((self.journal_id.type == 'sale' and not self.journal_id.l10n_ar_is_pos) or (
```
Comment (vbe-odoo) [inline addons/l10n_ar/models/account_move.py:43]: @zaoral update the condition here if POS always electronic

Comment (zaoral) [reply addons/l10n_ar/models/account_move.py]: Hi @vbe-odoo 
I have applied the changes but not sure if they are the right changes. I understood/applied the changes taking into account that is it a Pos System, do not take into account if it is electronic or not. Please let me know if you think we should take into account if it is electronic.
Wait your feedback

Best Regards

Comment (vbe-odoo) [reply addons/l10n_ar/models/account_move.py]: Hi @zaoral !
Dismiss my message about the electronic hehe! If l10n_ar_is_pos = True (selected) then it should work as a regular invoice (taking the number from AFIP and not being manual)
Cheers!

Comment (vbe-odoo) [reply addons/l10n_ar/models/account_move.py]: So it should work now, as you took into account the POS.
I can test! 🚀

Comment (vbe-odoo) [reply addons/l10n_ar/models/account_move.py]: Hello @zaoral can you help me do a rebase on this PR?
As there is an issue that my be solved with a rebase or if not if you can help me check what's happening.
So, when installing AR there is a conflict with some demos data DOCs and then it does not install the demo data, see the screenshot below:
image

Thanks!

File: addons/l10n_ar/models/account_journal.py [hunk 2]
```diff
@@ -112,16 +136,22 @@ def _check_afip_configurations(self):
                 _("You can not change the journal's configuration if it already has validated invoices") + ' ('
                 + ', '.join(invoices.mapped('journal_id').mapped('name')) + ')')
 
+    @api.constrains('l10n_ar_afip_pos_system')
+    def _check_afip_pos_system(self):
+        journals = self.filtered(
+            lambda x: x.l10n_ar_is_pos and x.type == 'purchase' and
+            self.l10n_ar_afip_pos_system not in ['II_IM', 'RLI_RLM', 'RAW_MAW'])
```
Comment (vin-odoo) [inline addons/l10n_ar/models/account_journal.py:143]:
```suggestion
            lambda j: j.l10n_ar_is_pos and j.type == 'purchase' and
            j.l10n_ar_afip_pos_system not in ['II_IM', 'RLI_RLM', 'RAW_MAW'])
```

File: addons/l10n_ar/models/account_journal.py [hunk 3]
```diff
@@ -112,16 +136,22 @@ def _check_afip_configurations(self):
...
+                ) % (x.l10n_ar_afip_pos_system, x.id) for x in journals]))
+
     @api.constrains('l10n_ar_afip_pos_number')
     def _check_afip_pos_number(self):
-        to_review = self.filtered(
-            lambda x: x.type == 'sale' and x.l10n_latam_use_documents and
-            x.company_id.account_fiscal_country_id.code == "AR")
-
-        if to_review.filtered(lambda x: x.l10n_ar_afip_pos_number == 0):
+        if self.filtered(lambda x: x.l10n_ar_is_pos and x.l10n_ar_afip_pos_number == 0):
```
Comment (vin-odoo) [inline addons/l10n_ar/models/account_journal.py:151]:
```suggestion
        if self.filtered(lambda j: j.l10n_ar_is_pos and j.l10n_ar_afip_pos_number == 0):
```
I don't mind using a single letter in the filter, but I prefer it to be the first of what we are iterating on. (Journals here)

File: addons/l10n_ar/models/account_journal.py [hunk 4]
```diff
@@ -112,16 +136,22 @@ def _check_afip_configurations(self):
...
             raise ValidationError(_('Please define an AFIP POS number'))
 
-        if to_review.filtered(lambda x: x.l10n_ar_afip_pos_number > 99999):
+        if self.filtered(lambda x: x.l10n_ar_is_pos and x.l10n_ar_afip_pos_number > 99999):
```
Comment (vin-odoo) [inline addons/l10n_ar/models/account_journal.py:154]:
```suggestion
        if self.filtered(lambda j: j.l10n_ar_is_pos and j.l10n_ar_afip_pos_number > 99999):
```

File: addons/l10n_ar/models/account_move.py [hunk 5]
```diff
@@ -99,10 +109,8 @@ def _get_l10n_latam_documents_domain(self):
         if self.journal_id.company_id.account_fiscal_country_id.code == "AR":
             letters = self.journal_id._get_journal_letter(counterpart_partner=self.partner_id.commercial_partner_id)
             domain += ['|', ('l10n_ar_letter', '=', False), ('l10n_ar_letter', 'in', letters)]
-            codes = self.journal_id._get_journal_codes()
-            if codes:
-                domain.append(('code', 'in', codes))
-            if self.move_type == 'in_refund':
+            domain += self.journal_id._get_journal_codes_domain()
```
Comment (vin-odoo) [inline addons/l10n_ar/models/account_move.py:112]: Would be better to replace this part and user expressions, such as `expression.AND`

**Review by vin-odoo — CHANGES_REQUESTED** [review]
Did a first tech review. I will probably let @jco-odoo check too as he is a lot more familiar with l10n_ar.

**Comment by zaoral** [comment]
Hi @vin-odoo 
Thank you for the review.
I have directly applied some or just suggestions and other ones will stay open mind while I review then more close it

Best Regards

File: addons/l10n_ar/models/account_move.py [hunk 6]
```diff
@@ -34,6 +34,25 @@ def _l10n_ar_get_document_number_parts(self, document_number, document_type_code
...
+        from there """
+        if self.country_code != 'AR':
+            return super()._is_manual_document_number()
+
+        # NOTE: There is a corner case where 2 sales documents can have the same number for the same DOC from a
+        # different vendor, in that case, the user can create a new Sales Liquido Producto Journal
+        return True if self.l10n_latam_use_documents and (
+            (self.journal_id.type == 'purchase' and not self.journal_id.l10n_ar_is_pos) or
+            (self.journal_id.type == 'sale' and not self.journal_id.l10n_ar_is_pos)
+        ) else False
```
Comment (jco-odoo) [inline addons/l10n_ar/models/account_move.py:54]:
```suggestion
        return self.l10n_latam_use_documents and (
            (self.journal_id.type == 'purchase' and not self.journal_id.l10n_ar_is_pos) or
            (self.journal_id.type == 'sale' and not self.journal_id.l10n_ar_is_pos)
        ) 
```

Comment (vbe-odoo) [reply addons/l10n_ar/models/account_move.py:54]: @zaoral ping

Comment (zaoral) [reply addons/l10n_ar/models/account_move.py:54]: push the last version with the simpler if, and remove the f/else condition

File: addons/l10n_ar/models/account_move.py (hunk 5 above)
Comment (vbe-odoo) [reply addons/l10n_ar/models/account_move.py:121]: @zaoral ping

File: addons/l10n_ar/models/account_move.py (hunk 6 above)
Comment (zaoral) [reply addons/l10n_ar/models/account_move.py:54]: add unit tests

**Comment by vbe-odoo** [comment]
@zaoral 
I've done the FUN test and now it works OK and looks good!

Also, all runbot tests - expect style. - are green! :🎉 

@jco-odoo this is ready for technical review.
Thanks!
cc' @fvz-odoo @ren-odoo

**Comment by zaoral** [comment]
Hi @vbe-odoo 

Thank you for the review.

I have updated the pr to the last changes we agreed on, resolved the pending runbot error for styling, and also added the unit test for liquido producto sales/purchase to ensure when we are asking the manual numbering or not taking into account manual numbering and is AFIP POS new field.

Is ready for your review @jco-odoo
both this one and https://github.com/odoo/enterprise/pull/36322

Best Regards

File: addons/l10n_ar/models/account_move.py (hunk 6 above)
Comment (zaoral) [reply addons/l10n_ar/models/account_move.py]: we simplify the method as we mentioned in the last call. and also added unit tests

File: addons/l10n_ar/models/account_move.py (hunk 5 above)
Comment (zaoral) [reply addons/l10n_ar/models/account_move.py]: we change the approach to use expression.AND

File: addons/l10n_ar/models/account_journal.py (hunk 4 above)
Comment (zaoral) [reply addons/l10n_ar/models/account_journal.py]: this one was also tested and applied

**Comment by vbe-odoo** [comment]
Thanks for the changes and the ping @zaoral 

I've done the FUN tests and LGTM, and this is ready for technical review @jco-odoo 
This PR has another one linked that is: https://github.com/odoo/enterprise/pull/36322
Thanks!
cc' @ren-odoo @fvz-odoo

File: addons/l10n_ar/models/account_journal.py [hunk 7]
```diff
@@ -9,14 +9,27 @@ class AccountJournal(models.Model):
...
+
+    @api.depends('country_code', 'type', 'l10n_latam_use_documents')
+    def _compute_l10n_ar_is_pos(self):
+        for journal in self:
+            journal.l10n_ar_is_pos = journal.country_code == 'AR' and journal.type == 'sale' and journal.l10n_latam_use_documents
+
+    @api.depends('l10n_ar_is_pos')
+    def _compute_l10n_ar_afip_pos_system(self):
+        for journal in self:
+            journal.l10n_ar_afip_pos_system = journal.l10n_ar_is_pos and journal.l10n_ar_afip_pos_system
```
Comment (jco-odoo) [inline addons/l10n_ar/models/account_journal.py:32]: This is a weird compute.  Why not an onchange to set values back?  Or is this the new style?

File: addons/l10n_ar/models/account_journal.py [hunk 8]
```diff
@@ -65,18 +78,16 @@ def _get_journal_letter(self, counterpart_partner=False):
...
             self.company_id.l10n_ar_afip_responsibility_type_id.code]
         if counterpart_partner:
-            counterpart_letters = letters_data['issued' if self.type == 'purchase' else 'received'].get(
+            counterpart_letters = letters_data['issued' if not self.l10n_ar_is_pos else 'received'].get(
                 counterpart_partner.l10n_ar_afip_responsibility_type_id.code, [])
             letters = list(set(letters) & set(counterpart_letters))
         return letters
 
-    def _get_journal_codes(self):
+    def _get_journal_codes_domain(self):
```
Comment (jco-odoo) [inline addons/l10n_ar/models/account_journal.py:89]: Why do we have a separate _get_journal_codes_domain and _get_codes_per_journal_type method then?

File: addons/l10n_ar/models/account_move.py [hunk 9]
```diff
@@ -99,10 +117,11 @@ def _get_l10n_latam_documents_domain(self):
...
-            codes = self.journal_id._get_journal_codes()
-            if codes:
-                domain.append(('code', 'in', codes))
-            if self.move_type == 'in_refund':
+            domain = expression.AND([
+                domain or [],
+                self.journal_id._get_journal_codes_domain(),
+            ])
+            if self.move_type in ['out_refund', 'in_refund']:
                 domain = ['|', ('code', 'in', self._get_l10n_ar_codes_used_for_inv_and_ref())] + domain
```
Comment (jco-odoo) [inline addons/l10n_ar/models/account_move.py:125]: so why not use expression.OR here then?

**Review by jco-odoo — COMMENTED** [review]
To check.

**Comment by jco-odoo** [comment]
@robodoo r+

**Comment by jco-odoo** [comment]
@zaoral Please check for the changes still afterwards, but this way, at least we have it.

**Pull request closed without merging**